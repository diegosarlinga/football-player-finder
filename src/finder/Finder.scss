.--loading {
  height: 100vh;
  overflow: hidden;
}