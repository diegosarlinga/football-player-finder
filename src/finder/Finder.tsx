import React from 'react';
import { connect } from 'react-redux';
import { LoadingIndicator, ErrorMessage, Content } from '../shared/components';
import Filter from './components/filter/Filter';
import List from './components/list/List';
import './Finder.scss';
import { AppState } from '../store';
import { fetchFootballPlayers, filterFootballPlayers } from './actions/actions';
import { FootballPlayer, FootballPlayerFilter } from './store/finder-state';
import { filteredFootballPlayerListSelector } from './store/selectors';

interface FinderProps {
  fetchFootballPlayers: () => void;
  filterFootballPlayers: (filter: FootballPlayerFilter) => void;
  isLoading: boolean;
  error: boolean;
  footballPlayers: Array<FootballPlayer>;
  positions: Array<string>;
  activeFilter: FootballPlayerFilter;
}

export class Finder extends React.Component<FinderProps> {

  componentDidMount() {
    this.props.fetchFootballPlayers();
  }

  renderContent() {
    return (
      <React.Fragment>
        <Filter
          positions={this.props.positions}
          activeFilter={this.props.activeFilter}
          onFilter={this.props.filterFootballPlayers}></Filter>
        <List footballPlayers={this.props.footballPlayers}></List>
      </React.Fragment>
    );
  }

  render() {
    const className = this.props.isLoading ? '--loading' : '';
    return (
      <Content title={'Football Player Finder'} className={className}>
        { this.props.isLoading && <LoadingIndicator/>}
        { this.props.error && <ErrorMessage/> }
        { !this.props.error && this.renderContent() }
      </Content>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  isLoading: state.finder.isLoading,
  error: state.finder.error,
  footballPlayers: filteredFootballPlayerListSelector(state),
  positions: state.finder.positions,
  activeFilter: state.finder.activeFilter,
});

const mapDispatchToProps = {
  fetchFootballPlayers: fetchFootballPlayers,
  filterFootballPlayers: filterFootballPlayers,
};

export default connect(mapStateToProps, mapDispatchToProps)(Finder);
