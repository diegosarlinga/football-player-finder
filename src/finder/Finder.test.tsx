import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Finder } from './Finder';

const defaultProps = {
  fetchFootballPlayers: jest.fn(),
  filterFootballPlayers: jest.fn(),
  isLoading: false,
  error: false,
  footballPlayers: [],
  positions: [],
  activeFilter: {}
}
describe('Finder component tests', () => {
  let wrapper: ShallowWrapper;

  describe('when rendered with default props', () => {
    beforeEach(() => {
      const props = {
        ...defaultProps,
      };
      wrapper = shallow(<Finder {...props}></Finder>)
    });

    it('should fetch players', () => {
      expect(defaultProps.fetchFootballPlayers).toHaveBeenCalledTimes(1);
    });
  });

  describe('when rendered with players and filter', () => {
    beforeEach(() => {
      const props = {
        ...defaultProps,
        positions: ['Pos 1', 'Pos 2'],
        activeFilter: {
          name: 'Pos',
          positions: 'Pos 1',
          age: 29
        },
        footballPlayers: [{
            contractUntil: "2020-06-30",
            dateOfBirth: "1997-02-02",
            jerseyNumber: 43,
            name: "Cameron Borthwick-Jackson",
            nationality: "England",
            position: "Left-Back",
          }, {
            contractUntil: "2020-06-30",
            dateOfBirth: "1997-02-02",
            jerseyNumber: 43,
            name: "Cameron Borthwick-Jackson",
            nationality: "England",
            position: "Left-Back",
        }]
      };
      wrapper = shallow(<Finder {...props}></Finder>)
    });

    it('should match normal state snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when rendered and is loading', () => {
    beforeEach(() => {
      const props = {
        ...defaultProps,
        isLoading: true,
      };
      wrapper = shallow(<Finder {...props}></Finder>)
    });

    it('should match loading state snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when rendered error state', () => {
    beforeEach(() => {
      const props = {
        ...defaultProps,
        error: true,
      };
      wrapper = shallow(<Finder {...props}></Finder>);
    });

    it('should match error state snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
