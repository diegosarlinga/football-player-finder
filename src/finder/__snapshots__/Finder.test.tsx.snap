// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Finder component tests when rendered and is loading should match loading state snapshot 1`] = `
<Content
  className="--loading"
  title="Football Player Finder"
>
  <LoadingIndicator />
  <Filter
    activeFilter={Object {}}
    onFilter={[MockFunction]}
    positions={Array []}
  />
  <List
    footballPlayers={Array []}
  />
</Content>
`;

exports[`Finder component tests when rendered error state should match error state snapshot 1`] = `
<Content
  className=""
  title="Football Player Finder"
>
  <ErrorMessage />
</Content>
`;

exports[`Finder component tests when rendered with players and filter should match normal state snapshot 1`] = `
<Content
  className=""
  title="Football Player Finder"
>
  <Filter
    activeFilter={
      Object {
        "age": 29,
        "name": "Pos",
        "positions": "Pos 1",
      }
    }
    onFilter={[MockFunction]}
    positions={
      Array [
        "Pos 1",
        "Pos 2",
      ]
    }
  />
  <List
    footballPlayers={
      Array [
        Object {
          "contractUntil": "2020-06-30",
          "dateOfBirth": "1997-02-02",
          "jerseyNumber": 43,
          "name": "Cameron Borthwick-Jackson",
          "nationality": "England",
          "position": "Left-Back",
        },
        Object {
          "contractUntil": "2020-06-30",
          "dateOfBirth": "1997-02-02",
          "jerseyNumber": 43,
          "name": "Cameron Borthwick-Jackson",
          "nationality": "England",
          "position": "Left-Back",
        },
      ]
    }
  />
</Content>
`;
